frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+
+			// enqueue them.
+			Queue::<T>::insert(ledger.stash, maybe_pool_id);
+			Ok(())
+		}
+
+		/// Deregister oneself from the fast-unstake (and possibly joining a pool).
+		///
+		/// This is useful if one is registered, they are still waiting, and they change their mind.
+		///
+		/// Note that the associated stash is still fully unbonded and chilled as a consequence of
+		/// calling `register_fast_unstake`. This should probably be followed by a call to
+		/// `Staking::rebond`.
+		#[pallet::weight(<T as Config>::WeightInfo::deregister())]
+		pub fn deregister(origin: OriginFor<T>) -> DispatchResult {
+			let ctrl = ensure_signed(origin)?;
+			let stash = pallet_staking::Ledger::<T>::get(&ctrl)
+				.map(|l| l.stash)
+				.ok_or(Error::<T>::NotController)?;
+			ensure!(Queue::<T>::contains_key(&stash), Error::<T>::NotQueued);
+			ensure!(
+				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != stash),
+				Error::<T>::AlreadyHead
+			);
+			Queue::<T>::remove(stash);
+			Ok(())
+		}
+
+		/// Control the operation of this pallet.
+		///
+		/// Dispatch origin must be signed by the [`Config::ControlOrigin`].
+		#[pallet::weight(<T as Config>::WeightInfo::control())]
+		pub fn control(origin: OriginFor<T>, eras_to_check: EraIndex) -> DispatchResult {
+			let _ = T::ControlOrigin::ensure_origin(origin)?;
+			ErasToCheckPerBlock::<T>::put(eras_to_check);
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// process up to `remaining_weight`.
+		///
+		/// Returns the actual weight consumed.
+		///
+		/// Written for readability in mind, not efficiency. For example:
+		///
+		/// 1. We assume this is only ever called once per `on_idle`. This is because we know that
+		/// in all use cases, even a single nominator cannot be unbonded in a single call. Multiple
+		/// calls to this function are thus not needed. 2. We will only mark a staker
+		#[cfg_attr(feature = "runtime-benchmarks", allow(unused_variables))]
+		pub(crate) fn process_head(remaining_weight: Weight) -> Weight {
+			let eras_to_check_per_block = ErasToCheckPerBlock::<T>::get();
+			if eras_to_check_per_block.is_zero() {
+				return T::DbWeight::get().reads(1)
+			}
+
+			// NOTE: here we're assuming that the number of validators has only ever increased,
+			// meaning that the number of exposures to check is either this per era, or less.
+			let validator_count = pallet_staking::ValidatorCount::<T>::get();
+
+			// determine the number of eras to check. This is based on both `ErasToCheckPerBlock`
+			// and `remaining_weight` passed on to us from the runtime executive.
+			let final_eras_to_check = {
+				let worse_weight = |v, u| {
+					<T as Config>::WeightInfo::on_idle_check(v * u)
+						.max(<T as Config>::WeightInfo::on_idle_unstake())
+				};
+				let mut try_eras_to_check = eras_to_check_per_block;
+				while worse_weight(validator_count, try_eras_to_check).all_gt(remaining_weight) {
```

Comment:
```suggestion
				while worse_weight(validator_count, try_eras_to_check).any_gt(remaining_weight) {
```

Because we want to ensure that `try_eras_to_check` has to have a worse weight strictly less than the remaining weight?